src/hotspot/share/gc/shared/slidingForwarding.inline.hpp

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright (c) 2021, Red Hat, Inc. All rights reserved.
+ * Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#ifndef SHARE_GC_SHARED_SLIDINGFORWARDING_INLINE_HPP
+#define SHARE_GC_SHARED_SLIDINGFORWARDING_INLINE_HPP
+
+#include "gc/shared/gc_globals.hpp"
+#include "gc/shared/slidingForwarding.hpp"
+#include "oops/markWord.hpp"
+#include "oops/oop.inline.hpp"
+#include "utilities/macros.hpp"
+
+inline bool SlidingForwarding::is_forwarded(oop obj) {
+  return obj->is_forwarded();
+}
+
+inline bool SlidingForwarding::is_not_forwarded(oop obj) {
+  return !obj->is_forwarded();
+}
+
+uint SlidingForwarding::region_index_containing(HeapWord* addr) {
+  uint index = static_cast<uint>(pointer_delta(addr, _heap_start) >> _region_size_words_shift);
+  assert(index < _num_regions, "Region index is in bounds: " PTR_FORMAT, p2i(addr));
+  return index;
+}
+
+uintptr_t SlidingForwarding::encode_forwarding(HeapWord* from, HeapWord* to) {
+  uint from_reg_idx = region_index_containing(from);
+  uint to_reg_idx = region_index_containing(to);
+
+  HeapWord* to_region_base = _heap_start + to_reg_idx * _region_size_words;
+  uint base_idx = from_reg_idx * NUM_TARGET_REGIONS;
+
+  bool alt_region = false;
+  if (_bases_table[base_idx] == UNUSED_BASE) {
+    // Primary is free
+    _bases_table[base_idx] = to_region_base;
+  } else if (_bases_table[base_idx] == to_region_base) {
```

Comment:
This probably won't help at all with performance, but I would kind of put the checks for the common cases where the table values are set (particularly the first one) first (I may be wrong about whether this is possible).
The `UNUSED_BASE` values in the tables will be encountered exactly once...

Reply:
I believe we can safely swap the UNUSED with the primary check. I'll do that.